backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java

```diff
@@ -629,50 +695,87 @@ private SyncToParentStatus deriveSyncToParentStatus(
+    @AllArgsConstructor // needed for @With
+    @SuppressWarnings("interning:not.interned") // to allow for `==` comparison in Lombok-generated `withSubentries` method
+    @UsesObjectEquals
+    private static class MyBranchLayoutEntry implements IBranchLayoutEntry {
+      @Getter
+      private final String name;
+
+      @Getter
+      @With
+      private List<IBranchLayoutEntry> subentries;
+
+      private @NotOnlyInitialized MyBranchLayoutEntry root;
+
+      MyBranchLayoutEntry(String name) {
+        this.name = name;
+        this.subentries = List.empty();
+        this.root = this;
       }
 
-      String remoteTrackingBranchName = branch.getRemoteTrackingBranch().map(rtb -> rtb.getName()).getOrNull();
+      void attachUnder(MyBranchLayoutEntry newRoot) {
+        root = newRoot;
+      }
 
-      var commitAndContainingBranches = gitCoreRepository
-          .ancestorsOf(branch.derivePointedCommit())
-          .map(commit -> {
-            var containingManagedBranches = deriveBranchesContainingGivenCommitInReflog()
-                .getOrElse(commit.getHash(), List.empty())
-                .filter(candidateBranchName -> !candidateBranchName.equals(branch.getName())
-                    && !candidateBranchName.equals(remoteTrackingBranchName)
-            // We demand that the candidate branch is already managed.
-                    && managedBranchNames.contains(candidateBranchName));
-            return Tuple.of(commit, containingManagedBranches);
-          })
-          .find(ccbs -> ccbs._2.nonEmpty())
-          .getOrNull();
+      void appendSubentry(MyBranchLayoutEntry newSubentry) {
+        subentries = subentries.append(newSubentry);
+      }
 
-      if (commitAndContainingBranches != null) {
-        var commit = commitAndContainingBranches._1;
-        var containingBranches = commitAndContainingBranches._2.toList();
-        assert containingBranches.nonEmpty() : "containingBranches is empty";
+      MyBranchLayoutEntry getRoot() {
+        // Path compression to reduce the lookup time,
+        // see https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Find
+        if (root != this && root.root != root) {
+          root = root.getRoot();
+        }
+        return root;
+      }
 
-        String firstContainingBranchName = containingBranches.head();
-        LOG.debug(() -> "Commit ${commit} found in filtered reflog(s) " +
-            "of managed branch(es) ${containingBranches.mkString(\", \")}; " +
-            "returning ${firstContainingBranchName} as the inferred upstream for branch '${localBranchName}'");
-        return Option.some(firstContainingBranchName);
-      } else {
-        LOG.debug(() -> "Could not infer upstream for branch '${branch.getFullName()}'");
+      @Override
+      public Option<String> getCustomAnnotation() {
         return Option.none();
       }
     }
+
+    IGitMacheteRepository discoverGitMacheteRepository() throws GitMacheteException, GitCoreException {
```

Comment:
What about wrapping `GitCoreException` into `GitMacheteException` to declare only the higher-level one?

Reply:
Done